File: clients/dotnet/Vapolia.Analytics.Client.Tests/CleanTests.cs
using Vapolia.Analytics.Client;

namespace Vapolia.Analytics.Client.Tests;

[TestClass]
public class CleanTests
{
    [TestMethod]
    public void TextTrimsCapsAndStripsControlCharacters()
    {
        Assert.AreEqual("hello", Clean.Text("  hello  ", 64));
        Assert.AreEqual("ab", Clean.Text("a\0b", 64));
        Assert.AreEqual(64, Clean.Text(new string('x', 100), 64)!.Length);
        Assert.IsNull(Clean.Text("   ", 64));
        Assert.IsNull(Clean.Text(null, 64));
    }

    [TestMethod]
    public void CountryIsTwoAsciiLettersOrNothing()
    {
        Assert.AreEqual("FR", Clean.Country("fr"));
        Assert.AreEqual("FR", Clean.Country(" FR "));
        // "GERMANY" truncated would be Georgia: unknown beats wrong.
        Assert.IsNull(Clean.Country("GERMANY"));
        Assert.IsNull(Clean.Country("F"));
        Assert.IsNull(Clean.Country("F1"));
        Assert.IsNull(Clean.Country(null));
    }

    [TestMethod]
    public void InstallIdIsACanonicalNonNilGuid()
    {
        Assert.AreEqual(
            "11111111-0000-0000-0000-000011111111",
            Clean.InstallId("11111111-0000-0000-0000-000011111111"));
        Assert.AreEqual(
            "aabbccdd-0000-0000-0000-00000000000f",
            Clean.InstallId("AABBCCDD-0000-0000-0000-00000000000F"));
        Assert.IsNull(Clean.InstallId("00000000-0000-0000-0000-000000000000"));
        Assert.IsNull(Clean.InstallId("not-a-guid"));
        // Braces and the dash-less form are not what the collector's clients send.
        Assert.IsNull(Clean.InstallId("{11111111-0000-0000-0000-000011111111}"));
        Assert.IsNull(Clean.InstallId(null));
    }

    [TestMethod]
    public void PropsKeepCappedScalarsOnly()
    {
        var kept = Clean.Props(new Dictionary<string, object?>
        {
            ["mode"] = new string('x', Clean.MaxValueLength + 10),
            ["ok"] = true,
            ["moves"] = 34,
            ["ratio"] = 0.5,
            ["nan"] = double.NaN,
            ["infinite"] = double.PositiveInfinity,
            ["blank"] = "   ",
            ["nothing"] = null,
            ["nested"] = new Dictionary<string, object>(),
            ["list"] = new[] { 1, 2 },
        })!;

        Assert.AreEqual(PropValue.From(true), kept["ok"]);
        Assert.AreEqual(PropValue.From(34d), kept["moves"]);
        Assert.AreEqual(PropValue.From(0.5d), kept["ratio"]);
        Assert.AreEqual(Clean.MaxValueLength, kept["mode"].StringValue!.Length);
        Assert.IsFalse(kept.ContainsKey("nan"));
        Assert.IsFalse(kept.ContainsKey("infinite"));
        Assert.IsFalse(kept.ContainsKey("blank"));
        Assert.IsFalse(kept.ContainsKey("nothing"));
        Assert.IsFalse(kept.ContainsKey("nested"));
        Assert.IsFalse(kept.ContainsKey("list"));
    }

    [TestMethod]
    public void PropsAreCappedInCountInKeyOrder()
    {
        var props = new Dictionary<string, object?>();
        for (var index = 0; index <= 20; index++)
            props[$"p{index:D2}"] = index;

        var kept = Clean.Props(props)!;

        Assert.AreEqual(Clean.MaxPropsPerEvent, kept.Count);
        Assert.IsTrue(kept.ContainsKey("p00"));
        Assert.IsFalse(kept.ContainsKey("p12"));
    }

    [TestMethod]
    public void DeviceNormalizesAndRefusesExcludedCountries()
    {
        var clean = new Device { Country = "fr" }.Cleaned()!;

        Assert.AreEqual("FR", clean.Country);

        Assert.IsNull(new Device { Country = "KR" }.Cleaned(["KR"]));
        Assert.IsNull(new Device { Country = "kr" }.Cleaned(["KR"]));
        Assert.AreEqual("KR", new Device { Country = "KR" }.Cleaned()!.Country, "no country is excluded unless the app lists it");
    }
}

File: clients/swift/Tests/VapoliaAnalyticsTests/CleanTests.swift
import XCTest
@testable import VapoliaAnalytics

final class CleanTests: XCTestCase {

    func testTextTrimsCapsAndStripsControlCharacters() {
        XCTAssertEqual(Clean.text("  hello  ", maxLength: 64), "hello")
        XCTAssertEqual(Clean.text("a\u{0}b", maxLength: 64), "ab")
        XCTAssertEqual(Clean.text(String(repeating: "x", count: 100), maxLength: 64)?.count, 64)
        XCTAssertNil(Clean.text("   ", maxLength: 64))
        XCTAssertNil(Clean.text(nil, maxLength: 64))
    }

    func testCountryIsTwoAsciiLettersOrNothing() {
        XCTAssertEqual(Clean.country("fr"), "FR")
        XCTAssertEqual(Clean.country(" FR "), "FR")
        // "GERMANY" truncated would be Georgia: unknown beats wrong.
        XCTAssertNil(Clean.country("GERMANY"))
        XCTAssertNil(Clean.country("F"))
        XCTAssertNil(Clean.country("F1"))
        XCTAssertNil(Clean.country(nil))
    }

    func testInstallIdIsACanonicalNonNilUuid() {
        XCTAssertEqual(
            Clean.installId("11111111-0000-0000-0000-000011111111"),
            "11111111-0000-0000-0000-000011111111"
        )
        XCTAssertEqual(
            Clean.installId("AABBCCDD-0000-0000-0000-00000000000F"),
            "aabbccdd-0000-0000-0000-00000000000f"
        )
        XCTAssertNil(Clean.installId("00000000-0000-0000-0000-000000000000"))
        XCTAssertNil(Clean.installId("not-a-uuid"))
        XCTAssertNil(Clean.installId("11111111000000000000000011111111"))
        XCTAssertNil(Clean.installId(nil))
    }

    func testPropsCapValues() {
        let kept = Clean.props([
            "mode": .string(String(repeating: "x", count: Limits.maxValueLength + 10)),
            "ok": true,
            "moves": 34,
            "nan": .number(.nan),
            "infinite": .number(.infinity),
            "blank": "   ",
        ])

        XCTAssertEqual(kept["ok"], .bool(true))
        XCTAssertEqual(kept["moves"], .number(34))
        if case .string(let mode)? = kept["mode"] {
            XCTAssertEqual(mode.count, Limits.maxValueLength)
        } else {
            XCTFail("mode should have survived, capped")
        }
        XCTAssertNil(kept["nan"])
        XCTAssertNil(kept["infinite"])
        XCTAssertNil(kept["blank"])
    }

    func testPropsAreCappedInCountInKeyOrder() {
        var props: [String: PropValue] = [:]
        for index in 0...20 {
            props[String(format: "p%02d", index)] = .number(Double(index))
        }

        let kept = Clean.props(props)

        XCTAssertEqual(kept.count, Limits.maxPropsPerEvent)
        XCTAssertEqual(kept["p00"], .number(0))
        XCTAssertNil(kept["p12"])
    }

    func testDeviceNormalizesAndRefusesExcludedCountries() throws {
        let clean = try XCTUnwrap(Device(country: "fr").cleaned())
        XCTAssertEqual(clean.country, "FR")

        XCTAssertNil(Device(country: "KR").cleaned(excluding: ["KR"]))
        XCTAssertNil(Device(country: "kr").cleaned(excluding: ["KR"]))
        XCTAssertEqual(Device(country: "KR").cleaned()?.country, "KR")
    }

    func testPropKeysAreCleanedLikeValues() {
        let kept = Clean.props([
            String(repeating: "k", count: Limits.maxValueLength + 10): .number(1),
            " mo\u{0}de ": .string("x"),
            "   ": .number(2),
        ])

        XCTAssertEqual(kept.count, 2)
        XCTAssertNotNil(kept[String(repeating: "k", count: Limits.maxValueLength)])
        XCTAssertEqual(kept["mode"], .string("x"))
    }
}
